numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
Answering one query: `__array_ufunc__ = None` is only meant to signal that that particular class cannot handle any `ufunc`. Within a forward binary operator, to a class that can handle ufuncs but does not know about that particular class, it should lead to `NotImplemented`, not to a `TypeError`.